### PR TITLE
Refactor vehicle lighting [RDY]

### LINKED ISSUE #13226: vehicle mounted atomic lamps
Now that atomic lamps can be covered by default, shouldn't vehicle mounted lamps be toggleable?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25544064-vehicle-mounted-atomic-lamps?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mugling**
Vehicle lights are now be toggled independently via their `enabled` field. This support is extended to include covering atomic lights so as to fix #13226.

Part of a series of refactors aimed at improving handling of vehicle power consumption.

### CODE REVIEW

**Comment by mugling** [comment]
Done

File: src/game.cpp [hunk 1]
```diff
@@ -1225,14 +1225,18 @@ bool game::cleanup_at_end()
     return true;
 }
 
-static int veh_lumi(vehicle *veh)
+static int veh_lumi( vehicle &veh )
```
Comment (Coolthulhu) [inline src/game.cpp]: Couldn't it also be `const`?

**Comment by Coolthulhu** [comment]
Will conflict with #17186

File: src/game.cpp (hunk 1 above)
Comment (mugling) [inline src/game.cpp]: Yes, but that requires defining a `const` variant of `vehicle::lights`.

What does that code section actually do - its repeated in `lightmap.cpp` and refactored it similarly but don't really understand it's significance.

**Comment by mugling** [comment]
> Will conflict with #17186

Yes, the predecessor to it was stalled but it looks like it's been recently updated. It's probably easier to rebase that PR against this one but I can go with either.

File: src/game.cpp (hunk 1 above)
Comment (mugling) [inline src/game.cpp]: Yes, but that requires defining a `const` variant of `vehicle::lights`.

What does that code section actually do - its repeated in `lightmap.cpp` and refactored it similarly but don't really understand it's significance.

EDIT: I'd rather avoid extensive changes if possible as this PR is relatively easy to test as-is

**Comment by mugling** [comment]
@Coolthulhu if I rebase is there time for this to go in your merge-test. I want to proceed with more changes to vehicle power that are dependent upon this PR...

**Comment by mugling** [comment]
Resolved merge conflict with turret refactor

**Comment by Coolthulhu** [comment]
Vehicles start with all lights on. This is rather annoying.
Another annoying thing is having to disable each light separately. There isn't a "SHUT DOWN EVERYTHING" button.

Red/blue lights, like on police car, don't work at all.

**Comment by mugling** [comment]
All changes

**Comment by Coolthulhu** [comment]
I think it's this one:
Newly created vehicles report not having an engine. I assume it's because they engine is disabled.
There is no way to enable it because the engine control menu doesn't appear until you have more than one engine.

**Comment by mugling** [comment]
@Coolthulhu yes, will look into

**Comment by Coolthulhu** [comment]
Having to manually turn on all the lights is a chore as it requires a lot of keypresses just to turn on the headlights.

It should be more like turrets: one global (per-vehicle) option and then individual config that allows customizing which lights we want on and which off.

**Comment by mugling** [comment]
I'm thinking about an extension supporting assigning lighting to circuits. It's a shame not to support arbitrary combinations of lights. This is a little beyond scope of this PR so I'd like to tackle it in the next

**Comment by Coolthulhu** [comment]
> This is a little beyond scope of this PR so I'd like to tackle it in the next

Having to manually turn on all lights is too much of a regression for me to allow it.
If this is to be merged, it needs at least a hack to make it usable. For example, bringing the old `lights_on` back, this time handling all lights at once (all enabled lights).

**Comment by mugling** [comment]
@Coolthulhu toggle headlights or toggle all lights?

**Comment by Coolthulhu** [comment]
Toggle all lights we had on last time.
Basically, we want 2 light states: all lights off, selected lights on. It must be able to switch between those with a minimal number of keypresses.

**Comment by mugling** [comment]
Have pushed fix for engine starting and toggle option for headlights

**Comment by Coolthulhu** [comment]
The "turn all lights off" hack has the same keybind as control lights.

**Comment by mugling** [comment]
Fixed

**Comment by Coolthulhu** [comment]
The light controls are unwieldy at the moment, so restoring usability is a priority here.
Specifically, restoring last light state - that's the most useful option regarding lights.
Circuits would be much less useful than restoring last light state. In fact, they would be less useful than the old functionality of per-type light config.

**Pull request merged**